### DISCUSSION TITLE
How to add this concentration function

### DISCUSSION THREAD

**Discussion by GuoChi-Li** [idaholab/moose#25252] [Q&A Modules: Phase field] [answered]
Hello MOOSE experts,

### Problem statement
I have a 2D phase-field problem, and I want to add this concentration function.
$$\frac{\partial c}{\partial \tilde{t}}= \nabla{\cdot\tilde{D}}[\nabla{c}+\frac{(1-k)c}{(1-\phi+k\phi)}\nabla{\phi}] \qquad (Eq. 1)$$
, where $D_s,D_l,\tau,l_0^2, k$ are constants; $\phi, c$ are variables; $\tilde{t}$ is time; $\tilde{D}$ is shown as follows
$$\tilde{D}=(D_s+\frac{(D_l-Ds)(1-\phi)}{(1-\phi+k\phi)})\frac{\tau}{l_0^2}$$

By doing some manipulation, I have this function (If this is not right, please let me know)
$$\frac{\partial c}{\partial \tilde{t}}= \nabla\tilde{D}\nabla{c}+\tilde{D}\nabla^{2}{c}+\nabla{\cdot}(\frac{\tilde{D}(1-k)c}{(1-\phi+k\phi)})\nabla{\phi}+(\frac{\tilde{D}(1-k)c}{(1-\phi+k\phi)})\nabla^{2}{\phi} \qquad (Eq. 2)$$
Let $$Dp=\frac{\tilde{D}(1-k)c}{(1-\phi+k\phi)}$$ and I get
$$\frac{\partial c}{\partial \tilde{t}}= \nabla\tilde{D}\nabla{c}+\tilde{D}\nabla^{2}{c}+\nabla{D_p}\nabla{\phi}+D_p\nabla^{2}{\phi} \qquad (Eq. 3)$$

### In Materials
```
  [./Dba]
    # This is tilde D
    type = DerivativeParsedMaterial
    property_name = Dba
    coupled_variables = 'phi'
    material_property_names = 'Ds Dl  k tao l02'
    expression = '(Ds+(Dl-Ds)*(1-phi)/(1-phi+k*phi))*(tao/l02)'
    derivative_order = 2
    outputs = exodus
  [../]
  [./Dp]
    type = DerivativeParsedMaterial
    property_name = Dp
    coupled_variables = 'c phi'
    material_property_names = 'Dba  k'
    expression = '(Dba*(1-k)/(1-phi+k*phi))*c'
    derivative_order = 2
    outputs = exodus
  [../]
```
### In Kernel
```
  [./c_dot]
    type = TimeDerivative
    variable = c
  [../]
  [./gradDba_dot_gradC]
    type = MatDiffusion
    variable = c
    diffusivity = Dba
    args = phi
  [../]
  [./Dba_dot_laplaceC]
    # TODO
  [../]
  [./gradDp_dot_gradPhi]
    type = MatDiffusion
    variable = c
    diffusivity = Dp
    args = 'c phi' # for diffusivity
    v = 'phi' # Coupled concentration variable for the kernel to operate on 
  [../]
  [./Dp_dot_laplacePhi]
    # TODO
  [../]
```
### Question
I don't know how to add **Dba_dot_laplaceC** and **Dp_dot_laplacePhi**. I have tried **FunctionDiffusion**, but it does not work. I think **FunctionDiffusion** can not work with variables. Is there any existing method I can use to add these? Or should I code a new one? Any comments are welcome.

Thank you.
Guo-Chi

**Comment by GiudGiud** [comment]
Hello

These terms are a little problematic here. You cant just do integration by parts / divergence theorem anymore because you dont have the divergence on the outside of the $D_p \nabla^2 c$ term for example.

You can make a new kernel to represent them.
You can use adCoupledSecond to obtain the second derivative.

You ll need c and phi to be at least second order so they dont have a 0 laplacian. 

The classic solution here though is I think to use the very first form of your equation and use simply 2 kernels, one MatDiffusion for c and a custom one for eta. 

Guillaume

**Reply by GuoChi-Li** [reply]
Hello

Thank you for the explanation. I find I have made some TYPOS in the code in Kernel. All the eta should be phi.

I have the following idea. 
$$\frac{\partial c}{\partial \tilde{t}}= \nabla{\cdot\tilde{D}}[\nabla{c}+\frac{(1-k)c}{(1-\phi+k\phi)}\nabla{\phi}] \qquad (Eq. 1)$$

$$\frac{\partial c}{\partial \tilde{t}}= \nabla{\cdot}[\tilde{D}\nabla{c}+\tilde{D}\frac{(1-k)c}{(1-\phi+k\phi)}\nabla{\phi}]$$

$$\frac{\partial c}{\partial \tilde{t}}= \nabla{\cdot}(\tilde{D}\nabla{c})+\nabla{\cdot}(\frac{\tilde{D}(1-k)c}{(1-\phi+k\phi)}\nabla{\phi})$$
Also, Let $$Dp=\frac{\tilde{D}(1-k)c}{(1-\phi+k\phi)}$$ and I get

$$\frac{\partial c}{\partial \tilde{t}}= \nabla{\cdot}(\tilde{D}\nabla{c})+\nabla{\cdot}(D_{p}\nabla{\phi}) \qquad (Eq. 4)$$

Maybe I can use two **MatDiffusion** kernels to implement now?

**Reply by GiudGiud** [reply, marked as answer]
Hello
Yes both these terms can be done with MatDiffusion and a DerivativeParsedMaterial for the Diffusion coefficients

Guillaume

**Reply by GuoChi-Li** [reply]
Thank you.
